prog/dftb+/lib_dftb/h5correction.F90

```diff
@@ -0,0 +1,243 @@
+
+    ! Local vars
+    integer :: iSp
+
+    ! Save system information used by H5
+    this%nSpecies = nType
+    this%speciesName = typeNames
+
+    ! Save H5 parameters
+    this%rScale = r
+    this%wScale = w
+    this%elementPara = elePara
+
+    ! Initialize the parameters
+    ! If value of the parameter is -1.0, it was not read from the input,
+    ! and default value will be used
+    if (this%rScale == -1.0_dp) then
+            this%rScale = 0.714_dp
+    end if
+
+    if (this%wScale == -1.0_dp) then
+            this%wScale = 0.25_dp
+    end if
+
+    do iSp = 1, this%nSpecies
+      if (this%elementPara(iSp) == -1.0_dp) then
+        ! Default values taken from the paper for O,N and S, zero for other elements
+        if (this%speciesName(iSp) == "O") then
+          this%elementPara(iSp) =  0.06_dp
+        else if (this%speciesName(iSp) == "N") then
+          this%elementPara(iSp) = 0.18_dp
+        else if (this%speciesName(iSp) == "S") then
+          this%elementPara(iSp) =  0.21_dp
+        else
+          this%elementPara(iSp) = 0.0_dp
+        end if
+      end if
+    end do
+
+  end subroutine H5Corr_init
+
+  !> Print all the parameters for debugging
+  subroutine printH5Setup(this)
+    ! Arguments
+    class(H5Corr), intent(in) :: this
+    ! Local variables
+    integer :: iSp1
+
+     write(37,*) "H5 setup:"
+     write(37,*) "   rScale = ", this%rScale
+     write(37,*) "   wScale = ", this%wScale
+     write(37,*) "H5 species parameters:"
+     do iSp1 = 1, this%nSpecies
+       write(37,*) "   ", this%speciesName(iSp1), " = ", this%elementPara(iSp1)
+     end do
+  end subroutine printH5Setup
+
+  !> Get H5 parameters for a pair of species
+  subroutine getParams(this, iSp1, iSp2, applyCorrection, h5Scaling, sumVDW)
+    !> Returns a flag whether the correction is applied to the pair,
+    !> and the species-specific parameters
+
+    ! Arguments
+    class(H5Corr), intent(in) :: this
+    !> Input: specie names
+    integer, intent(in) :: iSp1, iSp2
+    !> Output: flag indicating if the correction is applied to that pair
+    logical, intent(out) :: applyCorrection
+    !> Output: pair-specific scaling factor
+    real(dp), intent(out) :: h5Scaling
+    !> Output: Sum of vdW radii of the pair
+    real(dp), intent(out) :: sumVDW
+
+    ! Local variables
+    character(mc) :: spName1, spName2
+    integer :: iSpHeavy
+    character(mc) :: spNameHeavy
+
+    spName1 = this%speciesName(iSp1)
+    spName2 = this%speciesName(iSp2)
+    applyCorrection = .false.
+
+    ! If there is one hydrogen and one other atom, save the heavy atom,
+    ! otherwise return with applyCorrection = false
+    if (spName1 .ne. "H" .and. spName2 == "H") then
+      iSpHeavy = iSp1
+      spNameHeavy = spName1
+    else if (spName2 .ne. "H" .and. spName1 == "H") then
+      iSpHeavy = iSp2
+      spNameHeavy = spName2
+    else
+      return
+    end if
+   
+    ! For each species the correction is applied to,
```

Comment:
What is the fall back for other species outside this set?

Reply:
It returns applyCorrection=.false. unless there exist parameters for the pair. I'm adding comment on that.